ddl/partition.go

```diff
@@ -163,32 +163,114 @@ func checkPartitionNameUnique(tbInfo *model.TableInfo, pi *model.PartitionInfo)
 	return nil
 }
 
+// For partition tables, mysql do not support Constant, random or timezone-dependent expressions
+// Based on mysql code to check whether field is valid, every time related type has check_valid_arguments_processor function.
+// See https://github.com/mysql/mysql-server/blob/5.7/sql/item_timefunc
+type isTimezoneDependentProcessor func(columns []*expression.Column) bool
+
+// See https://github.com/mysql/mysql-server/blob/5.7/sql/item_func.h#L387
+func hasTimestampField(columns []*expression.Column) bool {
+	for _, c := range columns {
+		if c.GetType().Tp == mysql.TypeTimestamp {
+			return true
+		}
+	}
+
+	return false
+}
+
+// See https://github.com/mysql/mysql-server/blob/5.7/sql/item_func.h#L399
+func hasDateField(columns []*expression.Column) bool {
+	for _, c := range columns {
+		if c.GetType().Tp == mysql.TypeDate || c.GetType().Tp == mysql.TypeDatetime {
+			return true
+		}
+	}
+
+	return false
+}
+
+// See https://github.com/mysql/mysql-server/blob/5.7/sql/item_func.h#L412
+func hasTimeField(columns []*expression.Column) bool {
+	for _, c := range columns {
+		if c.GetType().Tp == mysql.TypeDatetime {
+			return true
+		}
+	}
+
+	return false
+}
+
+// We assume the result of any function that has a TIMESTAMP argument to be
+// timezone-dependent, since a TIMESTAMP value in both numeric and string
+// contexts is interpreted according to the current timezone.
+// The only exception is UNIX_TIMESTAMP() which returns the internal
+// representation of a TIMESTAMP argument verbatim, and thus does not depend on
+// the timezone.
+// See https://github.com/mysql/mysql-server/blob/5.7/sql/item_func.h#L445
+func defaultTimezoneDependent(columns []*expression.Column) bool {
+	return !hasTimestampField(columns)
+}
+
+var (
+	funcCallExprMap = map[string]isTimezoneDependentProcessor{
```

Comment:
The name `funcCallExprMap` is too generic, we can't infer what it does by the name.